### PR TITLE
Fix and enable deprecation warnings

### LINKED ISSUE #3473: [arm] M1 build from source with new BUILD_GUI set ON, Filament error
**Description**
I am able to build natively on apple's M1 chip with `BUILD_GUI` set to `OFF`. However when I switch `BUILD_GUI` to `ON`, I get errors related to filament at the end of the build. 

In both cases, I have `BUILD_FILAMENT_FROM_SOURCE` set to `ON`, and in the `filament_build.cmake` file, I have these options set like this 
```
set(FILAMENT_GIT_REPOSITORY "https://github.com/google/filament.git")
set(FILAMENT_GIT_TAG "main")
```

**To Reproduce**
1. git clone --recursive https://github.com/intel-isl/Open3D
2. Following these steps #2675
3. Set `option(BUILD_GUI                  "Builds new GUI"                           ON )` in `CMakeLists.txt`
4. See error after the `make install-pip-package -j8` command (end of build)

**Expected behavior**
Finish build .. ?

**Screenshots**
Screen Shot 2021-05-19 at 1 24 13 PM

```
/Users/mostafathereal/Open3D/cpp/open3d/visualization/rendering/filament/FilamentCamera.cpp:80:36: error: too few arguments to function call, single argument 'entity' was not specified
    camera_ = engine_.createCamera();
              ~~~~~~~~~~~~~~~~~~~~ ^
/Users/mostafathereal/Open3D/build/filament-binaries/include/filament/Engine.h:383:13: note: 'createCamera' declared here
    Camera* createCamera(utils::Entity entity) noexcept;
            ^
/Users/mostafathereal/Open3D/cpp/open3d/visualization/rendering/filament/FilamentCamera.cpp:85:45: error: no matching member function for call to 'destroy'
FilamentCamera::~FilamentCamera() { engine_.destroy(camera_); }
                                    ~~~~~~~~^~~~~~~
```

**Environment:**
- Operating system: Big Sur 11.2.3
- Python version: Python 3.9.4 - conda env
- Open3D version: 0.12.0
- Compiler version (if built from source): Apple clang version 12.0.5 (clang-1205.0.22.9) Target: arm64-apple-darwin20.3.0
- CMake version: 3.20.2
- conda 4.10.1

Thanks!

### PR DESCRIPTION
**Pull request by stotko**
Suppressing deprecation warnings hides important API changes in 3rd party code that will eventually lead to compiler errors in the future once the 3rd party library removes deprecated functions.

- Re-enable `-Wdeprecated` warnings and fix them (visualization, ml, tests, pybind)
- Re-enable `-Wunused-result` warnings and fix them (examples)

Fixes #3473

---
This change is [Reviewable](https://reviewable.io/reviews/intel-isl/open3d/3527)

### CODE REVIEW

File: cpp/open3d/visualization/rendering/filament/FilamentCamera.h [hunk 1]
```diff
@@ -26,14 +26,15 @@
...
+#include <utils/Entity.h>
+
 #include "open3d/visualization/rendering/Camera.h"
 
 /// @cond
 namespace filament {
 class Camera;
 class Engine;
 }  // namespace filament
-/// @endcond
```
Comment (ssheorey) [inline cpp/open3d/visualization/rendering/filament/FilamentCamera.h:36]: Missing `@endcond`? This will cause doxygen to ignore the rest of this file.

**Review by ssheorey — COMMENTED** [review]
Thanks @stotko this is useful. There are some build errors in macOS and Windows.

File: cpp/open3d/visualization/rendering/filament/FilamentCamera.h (hunk 1 above)
Comment (stotko) [reply cpp/open3d/visualization/rendering/filament/FilamentCamera.h]: Oops, that was a mistake when cleaning up my first attempt. Fixed.

**Comment by stotko** [comment]
> Thanks @stotko this is useful. There are some build errors in macOS and Windows.

Should be hopefully fixed now. I will continue fixing any further occurring build issues. Once CI is happy, this is ready.

**Review by yxlao — APPROVED** [review]
> Reviewed 8 of 9 files at r1, 3 of 3 files at r2, 2 of 2 files at r3.
> *[Reviewable](https://reviewable.io/reviews/intel-isl/open3d/3527)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @errissa)

**Pull request merged**